chalk-ir/src/fold/boring_impls.rs

```diff
@@ -71,25 +71,26 @@ impl<I: Interner, TI: TargetInterner<I>> Fold<I, TI> for Parameter<I> {
     type Result = Parameter<TI>;
     fn fold_with(&self, folder: &mut dyn Folder<I, TI>, binders: usize) -> Fallible<Self::Result> {
         let data = self.data().fold_with(folder, binders)?;
-        Ok(Parameter::new(data))
+        Ok(Parameter::new(folder.target_interner(), data))
     }
 }
 
 impl<I: Interner, TI: TargetInterner<I>> Fold<I, TI> for Substitution<I> {
     type Result = Substitution<TI>;
     fn fold_with(&self, folder: &mut dyn Folder<I, TI>, binders: usize) -> Fallible<Self::Result> {
+        let folded: Vec<_> = self.iter().map(|p| p.fold_with(folder, binders)).collect();
```

Comment:
Had to collect here and blow to deal with lifetimes. `p.fold_with` takes a &mut ref to the folder and `target_interner()` takes a & ref to the folder.